dlt-database/migrations/0001-init_db.ts

```diff
@@ -0,0 +1,130 @@
+/* FIRST MIGRATION
+ *
+ * This migration is special since it takes into account that
+ * the database can be setup already but also may not be.
+ * Therefore you will find all `CREATE TABLE` statements with
+ * a `IF NOT EXISTS`, all `INSERT` with an `IGNORE` and in the
+ * downgrade function all `DROP TABLE` with a `IF EXISTS`.
+ * This ensures compatibility for existing or non-existing
+ * databases.
+ */
+
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
+export async function upgrade(queryFn: (query: string, values?: any[]) => Promise<Array<any>>) {
+  // write upgrade logic as parameter of queryFn
+  await queryFn(`
+    CREATE TABLE IF NOT EXISTS \`users\` (
+      \`id\` int(10) unsigned NOT NULL AUTO_INCREMENT,
+      \`gradido_id\` char(36) DEFAULT NULL,
+      \`derive1_pubkey\` binary(32) NOT NULL,
+      \`created_at\` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
+      \`confirmed_at\` datetime(3) DEFAULT NULL,
+      PRIMARY KEY (\`id\`),
+      UNIQUE KEY \`gradido_id\` (\`gradido_id\`),
+      UNIQUE KEY \`pubkey\` (\`pubkey\`)
+    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;`)
+
+  await queryFn(`
+    CREATE TABLE IF NOT EXISTS \`accounts\` (
+      \`id\` int(10) unsigned NOT NULL AUTO_INCREMENT,
+      \`user_id\` int(10) unsigned DEFAULT NULL,
+      \`derivation_index\` int(10) unsigned NOT NULL,
+      \`derive2_pubkey\` binary(32) NOT NULL,
+      \`type\` tinyint unsigned NOT NULL,
+      \`created_at\` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3),
+      \`confirmed_at\` datetime(3) DEFAULT NULL,
+      \`balance\` decimal(40,20) NOT NULL DEFAULT 0,
+      \`balance_date\` datetime(3) NOT NULL DEFAULT CURRENT_TIMESTAMP(3), 
```

Comment:
i think you could avoid differences in the balance and decay calculations when setting the balance_date to the given balance_date from the backend-transactions attribute.
If you set it like here with the current-timestamp you will get a difference which cause unnecessary effects in the verification later, right?

Reply:
When the account is created, there is no last transaction yet, so the starting value for balance_date is CURRENT_TIMESTAMP and for balance = 0.
This way I always have a valid balance set for my calculations.

For my calculations, I use the timestamp of the IOTA milestone, i.e. confirmed_at.